Move semantic tokens to LSP

allisonchou opened the pull request:
This change provides LSP with the information needed for token colorization. 

The full LSP spec for semantic tokens can be found [here](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_semanticTokens) and provides information regarding how we must return token information back to LSP.

To summarize the spec information: we need to return the semantic token data in an inlined integer array representation that encodes the following information: (1) token line number relative to the previous token, (2) token start character relative to the previous token, (3) token length, (4) token type, and (5) token modifiers. Each token will thus be represented by five numbers in the array.

There are 3 handlers involved with computing tokens:
- `SemanticTokensRangeHandler:` Calculates the tokens for a visible range. Only invoked on file open. This is beneficial since it may result in the faster rendering of tokens if the whole document tokens request is still processing.
- `SemanticTokensHandler:` Calculates the tokens for the entire file. Usually only invoked on file open, but may be invoked again if there is an error with computing the edits or if an edit we provide is too difficult for LSP to apply.
- `SemanticTokensEditHandler:` Calculates the edits in the file. An edit request is received every 500ms, or every time the user makes an edit in the file. We usually return edits back to LSP, but we may also return the whole set of tokens for the file if there's some synchronization error or other issue.

Notes:
- Tokens and document identifiers are cached between requests so that we can properly return edits. 
- There is still some classification work that needs to be done, mainly centering around synchronization and custom token types (parent issue is #46052). Due to the lack of custom token types, some token colorization does not currently match the colorization in local Roslyn.

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+                        throw new ArgumentException($"Token type {modifier} is not recognized.");
+                    }
+
+                    modifiers |= index + 1;
```
Using index and ORing it doesn't make sense to me, and adding one makes it more confusing. The enum ends up needing to be:

```
Value1
Value2
Ignored3
Value4
Ignored5
Ignored6
Ignored7
Value8
```

I think `TokenModifiers` should be a `[Flags]` enum so its clear, and the value should be used not the index.

Hopefully part of this comment makes sense, I don't know if I'm explaining myself right (or indeed, have understood the code right :P)

dibarbet replied:
+1 here, a flags enum seems appropriate

CyrusNajmabadi commented:
Can you clarify what you mean by: An edit request is received every 500ms

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/Initialize/InitializeHandler.cs:
```diff
@@ -58,6 +59,16 @@ public Task<LSP.InitializeResult> HandleRequestAsync(LSP.InitializeParams reques
+                    {
+                        DocumentProvider = new LSP.SemanticTokensDocumentProviderOptions { Edits = true },
+                        RangeProvider = true,
+                        Legend = new LSP.SemanticTokensLegend
```
what does 'legend' mean here?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,114 @@
+                }
+
+                index++;
+            }
```
this feels pretty inefficient.  a single edit at the start of a file seems like it would lead to N deletion/inserts. which would actually be more costly than just having the whole result produced.

Should we not be using a diff here between the two sequences, to find a minimal edit changing the old values to the new ones?

gundermanc replied:
Agree with this, the recommendation is to do a standard diff of the arrays and just return the minimal edits.

I don't recommend re-implementing diff from scratch if you can help it.

I'm not sure if Roslyn has a diff implementation lying about, but some I am aware of are:
- Google's DiffMatchPatch, ported to C#: https://www.nuget.org/packages?q=diffmatchpatch
- Editor's diff algorithm (unfortunately operates only on strings or TextBuffers): http://index/?query=ITextDifferencing
- Razor language server's diff abstraction

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            LSP.Range? range,
+            CancellationToken cancellationToken)
+        {
+            if (textDocument == null)
```
does this represent a bug in the caller?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            }
+
+            var classifiedSpans = await Classifier.GetClassifiedSpansAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (classifiedSpans == null)
```
i don't think this should ever happen.  does it?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            var lastLineNumber = 0;
+            var lastStartCharacter = 0;
+
+            var groupedSpans = classifiedSpans.GroupBy(s => s.TextSpan);
```
feels very weird that we're grouping anything.

allisonchou replied:
I added a clarifying comment in the latest updates that hopefully explains things more.
The GroupBy behavior is modeled after how the Omnisharp SemanticHighlightService works - https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/SemanticHighlight/SemanticHighlightService.cs#L53-L65

CyrusNajmabadi replied:
yeah, i would prefer we not do this.  Classifier.GetClassifiedSpansAsync will already report an ordered list of classification spans.  so you just need to handle the case of getting multiple spans in a row with teh same text span.

CyrusNajmabadi replied:
this comment applies.  the tokens will be in order already.  so i would just linearly process them.

CyrusNajmabadi replied:
note: the above is *very* expensive.  you're basically allocating multiple times *per* classification.

CyrusNajmabadi replied:
we still need to address thsi.  classified spans are sorted.  tehre is no need to group anything.  if we want to handle teh case of `(0, 5, identifier),(0, 5, static)` we can do so as we linearly traverse this list looking to see if multiple classified spans in a row refer to the same text span.

this is a blocker for me.

JoeRobich replied:
I agree with Cyrus. For OmniSharp the semantic feature in opt-in and we tell customers it is experimental.

Since these spans are in order, looping through them, processing the ClassificationType, and emitting a Token when the span changes would perform better.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+                ComputeNextToken(data, lines, ref lastLineNumber, ref lastStartCharacter, span);
+            }
+
+            return data.ToArray();
```
this sees unnecessary.  caller passes in the builder.  tehy can do the conversion to an array.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            IGrouping<TextSpan, ClassifiedSpan> span)
+        {
+            // Each semantic token is represented in LSP by five numbers:
+            //     1. Token line number, relative to the previous token
```
calling it 'line number' is a bit odd.  it seems like like a line-delta.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+
+            // 1. Token line number, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            data.Add(deltaLine);
```
assert this is >= 0.

CyrusNajmabadi replied:
instead of adding teh data one at a time, i would instead have a helper that returns the 5-tuple (or 5- outs).  then you just do all the mutation in one location.  i.e. it's easier to understand a purely functional piece of code that just computes information, rather than one that computes and mutates as it's going on.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            TextLineCollection lines,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            IGrouping<TextSpan, ClassifiedSpan> span)
```
calling this 'span' when tokens have a span as well is strange.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            // 3. Token length
+            data.Add(textSpan.Length);
+
+            var additiveResults = span.Where(s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
```
don't understand this.

allisonchou replied:
Just added a clarifying comment in the latest updates, but basically the line just tries to get the token's modifiers.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            var additiveResults = span.Where(s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
+
+            // 4. Token type - looked up in SemanticTokensLegend.tokenTypes.
+            var tokenTypeClassifiedSpan = span.Except(additiveResults).Single();
```
all this linq seems super expensive.  we optimized the heck out of classification to ensure low allocations and lightweight types.  it feels like we should be able to preserve this with a straight linear pass over the results.

allisonchou replied:
The behavior here is modeled after Omnisharp: https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/SemanticHighlight/SemanticHighlightService.cs#L53-L65

I can definitely see the concern with the LINQ queries. Currently ClassificationTypeNames.AdditiveTypeNames only contains one value, so it's relatively inexpensive, but I see how that can change in the future.

If you would be available to chat about this sometime, please let me know. I can definitely brainstorm more tomorrow, but I can't currently think of a way to do this with a straight linear pass (although my brain is a bit foggy rn).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHandler.cs:
```diff
@@ -0,0 +1,54 @@
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            var previousResultId = tokensCache.Tokens.ResultId == null ? 0 : int.Parse(tokensCache.Tokens.ResultId);
```
using 0 is worrying as it feels like it might be possible to confuse that with a real resultId of 0 passed in a caller.  can we use a nullable instead to properly convey when we do/don't have this information

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensTests.cs:
```diff
@@ -0,0 +1,58 @@
+            {
+                TextDocument = new LSP.TextDocumentIdentifier { Uri = caret.Uri }
+            };
+    }
```
do you have tests that validate the diff approach?

allisonchou replied:
This file deals with the diff approach: https://github.com/dotnet/roslyn/blob/eb3e158f7bfb2556cc541419c966ac12ec672fdd/src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs

And this test also tests the diff approach:
https://github.com/dotnet/roslyn/blob/eb3e158f7bfb2556cc541419c966ac12ec672fdd/src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensTests.cs#L53

Also, you were totally right with your earlier comment about the original diff logic being wrong. 😅

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,288 @@
+            var tokenTypeClassifiedSpan = span.Except(additiveResults).Single();
+            if (s_classificationTypeToSemanticTokenTypeMap.TryGetValue(tokenTypeClassifiedSpan.ClassificationType, out var tokenType))
+            {
+                var index = TokenTypes.IndexOf(t => t == tokenType);
```
The mapping of token name to int seems like something that should be defined in the LSP protocol code, at least for the base LSP types.  Otherwise every server has to maintain this separately.  Almost seems like int Enum values would be perfect here.

cc @gundermanc

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensTests.cs:
```diff
@@ -0,0 +1,58 @@
+    public class SemanticTokensTests : AbstractSemanticTokensTests
+    {
+        [Fact]
+        public async Task TestGetSemanticTokensAsync()
```
Can we add a test somewhere that invokes all the requests in the expected order (e.g. range, full document, edit1, edit2 and makes sure we get the expected result at each stage?

gundermanc replied:
Might also be worth checking that you can interleave full document requests in the middle there too. In the case of erroneous or hard to apply edits, the client may re-request the full document.

dibarbet commented on src/Features/LanguageServer/Protocol/AbstractRequestHandlerProvider.cs:
```diff
@@ -45,5 +46,26 @@ public AbstractRequestHandlerProvider(IEnumerable<Lazy<IRequestHandler, IRequest
+        /// Special handler for semantic tokens. Should only be needed to execute handlers that require caching, i.e.
+        /// <see cref="SemanticTokensHandler"/> and <see cref="SemanticTokensEditsHandler"/>.
+        /// </summary>
+        public Task<ResponseType> ExecuteSemanticTokensRequestAsync<RequestType, ResponseType>(
```
+1 here for creating a mef exported cache that we import in each handler.  Then we don't need special request handlers for semantic tokens + we could move the cache management out of InProcLanguageServer

dibarbet commented on src/VisualStudio/LiveShare/Impl/AbstractClassificationsHandler.cs:
```diff
@@ -1,56 +0,0 @@
-    /// Note, this must return object instead of ClassificationSpan b/c liveshare uses dynamic to convert handler results.
-    /// Unfortunately, ClassificationSpan is an internal type and cannot be defined in the external access layer.
-    /// </summary>
-    internal abstract class AbstractClassificationsHandler : ILspRequestHandler<ClassificationParams, object[], Solution>
```
All of the client side code for this should be deleted too
https://github.com/dotnet/roslyn/tree/master/src/VisualStudio/LiveShare/Impl/Client/Classification

Also the custom type - https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/LiveShare/Impl/CustomProtocol/Classifications.cs
and custom method name - https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/LiveShare/Impl/CustomProtocol/RoslynMethods.cs#L10

gundermanc commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -226,6 +230,34 @@ public Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(WorkspaceSymbolParams
+                SemanticTokensMethods.TextDocumentSemanticTokensName, semanticTokensParams, _tokensCache, _clientCapabilities,
+                _clientName, cancellationToken).ConfigureAwait(false);
+
+            _tokensCache.UpdateCache(semanticTokensParams.TextDocument, tokens);
```
>I would guess that the client would wait for edit requests to finish before sending more, but I'm not sure that is true for full document or range.

VS de facto only has one edits request in flight per document at a time, though if you have multiple documents open, we may have multiple out. That said, AFAIK, there's no explicit guarantee of this in the protocol spec.

>cc @davidwengier as well as, this seems like almost a 'write' request because we're updating the server cache.

IMO, the right way to handle this is to treat the cache as just a cache of the _n_ most recent items.. semantic tokens requests should be parallelizable but if there happens to be a cached version that matches, we return that. Then it's just a matter of properly synchronizing via atomic updates or a lock. Though VS doesn't do this.. there's no reason you couldn't have a client that might ask for changes on multiple versions right after one another, and in that case, a cache of _n_ items might be a useful way to think about it.. because scaling to that case just involves increasing the size of the cache.

You could choose to synchronize accesses to this cache by not respond to any other requests until this one completes, but this really shouldn't be a state-impacting request.

Once you have established the snapshot you are on, there's no reason not to start dispatching other requests. Such blocking behavior might significantly decrease throughput.

dibarbet replied:
I think Christian's point makes sense - the cache should handle a set of most recent values that we can drop as we get newer ones.  Then we just need to make sure we can synchronize updates to the cache properly.  Even if we handle the requests out of order slightly we still have both results in the cache which can be used by subsequent requests.

gundermanc commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -226,6 +230,34 @@ public Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(WorkspaceSymbolParams
+                _clientName, cancellationToken).ConfigureAwait(false);
+
+            _tokensCache.UpdateCache(semanticTokensEditsParams.TextDocument, tokens.SemanticTokens);
+            return tokens.SemanticTokensEdits ?? (SumType<SemanticTokens, SemanticTokensEdits>)tokens.SemanticTokens;
```
FWIW, SumType should have a constructor you can use instead of casting.

gundermanc commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,36 @@
+            Tokens = new LSP.SemanticTokens();
+        }
+
+        public void UpdateCache(LSP.TextDocumentIdentifier? document, LSP.SemanticTokens tokens)
```
Feel free to loop me into any conversations on this. I have few manual test cases that do a great job of finding synchronization issues.

davidwengier replied:
We're tracking this in https://github.com/dotnet/roslyn/issues/46051 and currently thinking about porting OmniSharp's [scheduler](https://github.com/OmniSharp/csharp-language-server-protocol/blob/a8e44c27cb4319bd0e6ffc77df75f9359752b563/src/JsonRpc/ProcessScheduler.cs#L78-L86) to Roslyn in order to use it, which is what Razor is using.

dibarbet commented:
@allisonchou just merged https://github.com/dotnet/roslyn/pull/46408 which should bring the feature branch up to date

allisonchou commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -226,6 +230,34 @@ public Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(WorkspaceSymbolParams
+            return tokens.SemanticTokensEdits ?? (SumType<SemanticTokens, SemanticTokensEdits>)tokens.SemanticTokens;
+        }
+
+        // Note: Since a range request is always received in conjunction with a whole document request, we don't need to cache range results.
```
Yep, I believe range requests are the first ones to be received, so the cache for the document isn't constructed yet.

allisonchou commented:
> Can you clarify what you mean by: An edit request is received every 500ms

Each time a user makes an edit or 500ms elapses, LSP will send an update request for the current semantic tokens using the edits request. If 500ms elapse, I believe LSP sends a request even if no edits to the document have been made.

allisonchou commented:
Thanks to everyone for the helpful feedback. 🎉 I believe all comments have been addressed either in the code or via a reply, other than the synchronization issues (tracked in #46052, should be done in a separate PR) .

The PR has changed a ton based on feedback. I would appreciate any more comments, although the logic/code has changed so much that I think it might be good to review it from the beginning again. Major changes:
- Cache is now imported via MEF
- Edit logic changed
- Added more tests
- Added enum type with flags for computing token modifiers
- Removed lots of unnecessary code

allisonchou commented on src/Workspaces/CoreTestUtilities/WorkspaceExtensions.cs:
```diff
@@ -30,7 +30,7 @@ public static void RemoveDocument(this Workspace workspace, DocumentId documentI
         }
 
-        public static void Updatedocument(this Workspace workspace, DocumentId documentId, SourceText newText)
+        public static void UpdateDocument(this Workspace workspace, DocumentId documentId, SourceText newText)
```
This doesn't have anything to do with the PR, just noticed that the casing was off.

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/IRequestHandler.cs:
```diff
@@ -6,6 +6,7 @@
 
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens;
```
This can presumably be removed now.

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,130 @@
+
+                // Return a non-null value only if the document's cache contains a token set with the resultId
+                // that the user is searching for.
+                return tokenSets.FirstOrDefault(t => t.ResultId != null && t.ResultId == resultId);
```
Nit: Perhaps a call to UpdateCacheAsync with a null ResultId should be a no-op, rather than cache something that can't be retrieved later.

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,141 @@
+        /// <summary>
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdits ComputeSemanticTokensEdits(
```
Does the size of edits matter? If I'm reading this right, for an insertion in the middle of a file this algorithm will produce two edits, rather than one? And the first of the two will be potentially very large.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,130 @@
+                NextResultId.TryGetValue(uri, out var resultId);
+                NextResultId[uri] = resultId + 1;
+                return resultId.ToString();
+            }
```
we coudl make this much simpler just by having an int/long, and doing an Interlocked.Increment on it.  bam. easy, unique ids.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,141 @@
+                // Case 1: Both cached and updated tokens have values at index, but the tokens don't match.
+                // We want to make the least number of edits possible, so we keep track of consecutive changes
+                // and report them as a singular edit.
+                if (cachedSemanticTokens[index] != updatedSemanticTokens[index])
```
yup, this algorithm still seems untenable.  consdier a simple case of 999 tokens before, and one new token added at the start after.    We start at index 0, and the tokens are different, so we add an edit. tehn we move index 1 and add an edit, and index 2 and so on.  so we end up reporting 999 edits in this case.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,141 @@
+        private static LSP.SemanticTokensEdits ComputeSemanticTokensEdits(
+            string resultId,
+            int[] cachedSemanticTokens,
+            int[] updatedSemanticTokens)
```
i recommend the names old/new.  this has the benefit of being easy to understand (Versus cached/updated), and is the same length, which makes thigns align nicely at times.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,141 @@
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdits ComputeSemanticTokensEdits(
+            string resultId,
```
feels weird to pass the resultId in.  it isn't used for computing the edits right?  instead, just have this be a helper method that literally just produces the edits.  teh caller can package the result+edits into the final result given to whatever called it.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,293 @@
+                [ClassificationTypeNames.XmlLiteralEntityReference] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralName] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralProcessingInstruction] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralText] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
```
we should have a tracking issue that tracks what we're doign to do here about having the rich set of colors for lsp.

allisonchou replied:
Just opened up an issue on our side here: https://github.com/dotnet/roslyn/issues/46582

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,293 @@
+
+            // 4. Token type - looked up in SemanticTokensLegend.tokenTypes (language server defined mapping
+            // from integer to LSP token types).
+            if (!s_classificationTypeToSemanticTokenTypeMap.TryGetValue(tokenTypeClassifiedSpan.ClassificationType, out var tokenTypeStr))
+            {
+                throw new NotSupportedException($"Classification type {tokenTypeClassifiedSpan.ClassificationType} is unsupported.");
+            }
+
+            var tokenTypeIndex = GetTokenTypeIndex(tokenTypeStr);
```
extract this bit into it's own helper.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,293 @@
+            var tokenTypeIndex = GetTokenTypeIndex(tokenTypeStr);
+
+            // 5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+            var modifierBits = TokenModifiers.None;
+            foreach (var currentModifier in modifiers)
+            {
+                if (!s_classificationTypeToSemanticTokenModifierMap.TryGetValue(currentModifier.ClassificationType, out var modifier))
+                {
+                    throw new NotSupportedException($"Classification type {currentModifier.ClassificationType} is unsupported.");
+                }
+
+                modifierBits |= modifier;
+            }
```
extract thsi bit as well.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,293 @@
+            }
+
+            return tokenTypeIndex;
+        }
```
this seems bad.  this is N^2 as we can have N tokesn that each do a N lookup here.   Seems like we can compute and cache the map from `.ClassificationType` to `index`.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,293 @@
+            var modifierBits = TokenModifiers.None;
+            foreach (var currentModifier in modifiers)
+            {
+                if (!s_classificationTypeToSemanticTokenModifierMap.TryGetValue(currentModifier.ClassificationType, out var modifier))
```
1. this seems like massive overkill given we have one modifier.
2. we'll throw an exception if anotehr modifier is ever added and we don't test lsp.  this feels not good.

allisonchou replied:
Would you suggest we just explicitly test for the 'static' modifier and remove all other logic? My only concern is that if another modifier is ever added, it will probably be very easy to miss updating LSP.
Another option could be to change the exception to a Debug assert?

CyrusNajmabadi replied:
yup, i'd check for the single case, and assert if we hit something else.  wherever the 'static' guy is defined, we can also put a comment saying that LSp woul dneed to be updated here.

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs:
```diff
@@ -0,0 +1,204 @@
+        /// Tests inserting a newline in the middle of the file.
+        /// </summary>
+        [Fact]
+        public async Task TestGetSemanticTokensEditsBasicAsync()
```
instead of a doc comment, remove that, and call this TestInsertingNewLineInMiddleOfFile.

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs:
```diff
@@ -0,0 +1,204 @@
+
+            var results = await RunGetSemanticTokensEditsAsync(workspace.CurrentSolution, caretLocation, previousResultId);
+            return results;
+        }
```
so, i don't like these tests ata ll :-/  we should chat about a better, but more involved, way of writing this stuff.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,212 @@
+        /// together by semantic token.
+        /// </summary>
+        private static SemanticToken[] ConvertToGroupedSemanticTokens(int[] tokens)
+        {
```
assert length is a multiple of 5.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,212 @@
+
+                return oldToken.DeltaLine == newToken.DeltaLine && oldToken.DeltaStartCharacter == newToken.DeltaStartCharacter &&
+                    oldToken.Length == newToken.Length && oldToken.TokenType == newToken.TokenType &&
+                    oldToken.TokenModifiers == newToken.TokenModifiers;
```
i would just implement == for SemanticToken.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,212 @@
+        /// <summary>
+        /// Stores the values that make up the LSP representation of an individual semantic token.
+        /// </summary>
+        private struct SemanticToken
```
readonly

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,212 @@
+                TokenModifiers = tokenModifiers;
+            }
+
+            public int[] ToArray()
```
don't do this.  this will allocate an array for every semantic token.  instead, just splat this into place in teh final array (many ways to do this).

CyrusNajmabadi replied:
(note: this may not apply.  i see how you're using ToArray now.

allisonchou replied:
Yes, I'm not sure if it applies since it's just converting our custom SemanticToken struct type back to it's original int array format.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,212 @@
+            public int DeltaStartCharacter { get; }
+            public int Length { get; }
+            public int TokenType { get; }
+            public int TokenModifiers { get; }
```
nit: just expose readonly fields.  the way people are enamored with props bothers me :)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,212 @@
+
+            // Since edits are reported relative to each other, we only care about an 'Update' EditKind
+            // if it's preceded by an insertion or deletion.
+            var adjustToken = false;
```
this part seems odd to me.  an update is just equivalent to a Delete+Insert.  So why do we need to do anything special.

gundermanc commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <summary>
+        /// Number of cached token sets we store per document. Must be >= 1.
+        /// </summary>
+        private readonly int _maxCachesPerDoc = 5;
```
Just FYI: in the local scenario, we'll almost always ask for a monotonically increasing PreviousResultId (the immediately previous semantic tokens result) but in the CodeSpaces and LiveShare scenarios, I've discovered that LiveShare is directly remoting semantic tokens requests from the client, so you could have up to 1 additional PreviousResultId version specified per attached guest.

gundermanc replied:
5 seems like a good initial cache size for handling this.

gundermanc commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <summary>
+        /// Maps a document URI to its n most recently cached token sets.
+        /// </summary>
+        private Dictionary<Uri, List<LSP.SemanticTokens>> Tokens { get; }
```
One thing to note about the cache being a MEF part is that these cached semantic tokens will be kept alive if the user closes the Roslyn workspace and opens a different project. Probably not a big deal, since this will eventually be OOP, but worth noting.

CyrusNajmabadi replied:
readonly?

gundermanc commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+                var initialEditIndex = editIndexes[i];
+                var initialEditKind = indexToEditKinds[initialEditIndex];
+
+                if (initialEditKind == EditKind.Update)
```
Can this be a switch statement or expression?

Also, might be nice to put the cases in helpers.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <summary>
+        /// Maps a LSP token type to its respective index recognized by LSP.
+        /// </summary>
+        public Dictionary<string, int> TokenTypesToIndex { get; }
```
readonly?

CyrusNajmabadi replied:
this dictionary doesn't seem to change.  i.e. it's entirely static.  shoudl store in a static field. c an be initialized with a static constructor.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to the Tokens or NextResultId dictionaries.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
```
general recommendation for locks+fields.  i would move everything that does not need the loc to the top of teh class.  then define the lock, and then make a `#region protected by _semaphore` that wraps the fields that should be tied to this.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+    /// Multiple token sets can be cached per document. The number of token sets cached
+    /// per document is determined by the _maxCachesPerDoc field.
+    /// </summary>
+    [Export(typeof(SemanticTokensCache)), Shared]
```
i'm not certain why this is an export.  why not just ahve this be instantiated by whatever other component needs to use this?

allisonchou replied:
The cache and its token are shared between the different semantic tokens handlers, so I don't think we can instantiate directly.

CyrusNajmabadi replied:
makes sense.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+                if (tokens.ResultId == null)
+                {
+                    return;
+                }
```
can move outside lock.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+
+                // Return a non-null value only if the document's cache contains a token set with the resultId
+                // that the user is searching for.
+                return tokenSets.FirstOrDefault(t => t.ResultId != null && t.ResultId == resultId);
```
`t.ResultId != null` is unnecessary.  if it's null it will never equal `resultId` (which you verified was non-null above).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently.
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to the Tokens or NextResultId dictionaries.
```
1. i don't see a NextResultId dictionary.  so this comment isn't valid.
2. even if you're referring to _nextResultId, you use an interlocked call to update that.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <summary>
+        /// Computes the mapping between a LSP token type and its respective index recognized by LSP.
+        /// </summary>
+        private static Dictionary<string, int> ComputeTokenTypesToIndex()
```
this is only called once, and from teh constructor.  just make a locl function there.  or inline this entirely and just have the constructor do this work.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+            ILspSolutionProvider solutionProvider,
+            SemanticTokensCache tokensCache) : base(solutionProvider)
+        {
+            _tokensCache = tokensCache;
```
no need to improt the cache.  just literally instantiate it on the field directly.

allisonchou replied:
(Addressed in https://github.com/dotnet/roslyn/pull/46407#discussion_r469612249).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+            // both for caching purposes and in order to have a baseline comparison when computing the edits.
+            var newSemanticTokens = await SemanticTokensHelpers.ComputeSemanticTokensAsync(
+                request.TextDocument, resultId, clientName, SolutionProvider, _tokensCache.TokenTypesToIndex,
+                range: null, cancellationToken).ConfigureAwait(false);
```
i would recommend *not* passing in resultId here, and instead having this just be concerned with computing the token-data.

CyrusNajmabadi replied:
that leaves resultId as a totally local piece of information.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+                request.TextDocument.Uri, request.PreviousResultId, cancellationToken).ConfigureAwait(false);
+            if (oldSemanticTokens == null)
+            {
+                return newSemanticTokens;
```
then, you just return `return new LSP.SemanticTokens { ResultId = resultId, Data = newData };` here.

CyrusNajmabadi replied:
also, in this codepath, yhou don't seem to be caching.  intentional?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+            // Getting the cached tokens for the document. If we don't have an applicable cached token set,
+            // we can't calculate edits, so we must return all semantic tokens instead.
+            var oldSemanticTokens = await _tokensCache.GetCachedTokensAsync(
+                request.TextDocument.Uri, request.PreviousResultId, cancellationToken).ConfigureAwait(false);
```
i would have this just return the data as well.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+            {
+                // We only care about EditKind.Insert and EditKind.Delete, since they encompass all
+                // changes to the document. All other EditKinds are ignored.
+                // However, we do use EditKind.Update to represent that an index contains both an
+                // insertion and deletion and thus can be combined.
```
```suggestion
                // However, we do use EditKind.Update to represent that an index contains both an
                // insertion and deletion and thus can be combined.
```

remove.  or update to say you're specifically ignoring this one, and why.

CyrusNajmabadi replied:
hrmm.. i'm probably wrong here.

allisonchou replied:
I realized the way I wrote out the logic was super confusing since I tried to re-use the existing EditKind enum to fulfill a different purpose. I pushed changes that use a new SemanticTokenEditKind enum instead that will hopefully make things clearer.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+                        break;
+                    case EditKind.Delete:
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        indexToEditKinds[edit.OldIndex] = editKindWithoutDelete == default ? EditKind.Delete : EditKind.Update;
```
i'm actually not certain i understand what's going on here.  you're taking multiple deletes at the same index and making it an Update?

allisonchou replied:
I refactored the code + added comments to try to explain this better, but basically this part is just seeing whether we can combine an insertion and deletion into a single edit.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,313 @@
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
```
todo: i need to review this.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHandler.cs:
```diff
@@ -0,0 +1,58 @@
+    [ExportLspMethod(LSP.SemanticTokensMethods.TextDocumentSemanticTokensName), Shared]
+    internal class SemanticTokensHandler : AbstractRequestHandler<LSP.SemanticTokensParams, LSP.SemanticTokens>
+    {
+        private readonly SemanticTokensCache _tokensCache;
```
oh interesting.  are you trying to share this cache among both multiple handlers?

CyrusNajmabadi replied:
if not, can a single handler handle multiple LSP methods?

allisonchou replied:
Yep! Trying to share among multiple handlers.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHandler.cs:
```diff
@@ -0,0 +1,58 @@
+                request.TextDocument, requestId, clientName, SolutionProvider, _tokensCache.TokenTypesToIndex,
+                range: null, cancellationToken).ConfigureAwait(false);
+
+            await _tokensCache.UpdateCacheAsync(request.TextDocument.Uri, tokens, cancellationToken).ConfigureAwait(false);
```
you always compute/cache tokens here (which is fine).  Should you also be doing this in the othe rhandler?  should ComputeSemanticTokensAsync just always cache the results?

allisonchou replied:
Both `SemanticTokensHandler` and `SemanticTokensEditHandler` use and update the cache.
`SemanticTokensRangeHandler` uses the cache (just to get the next resultId) but doesn't update the cache.
All three handlers call `ComputeSemanticTokensAsync`, so I don't think we can update the cache directly in this method.

CyrusNajmabadi replied:
thanks!

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+
+            // By default we calculate the tokens for the full document span, although the user 
+            // can pass in a range if they wish.
+            var textSpan = root.FullSpan;
+            if (range != null)
+            {
+                textSpan = ProtocolConversions.RangeToTextSpan(range, text);
+            }
```
```suggestion
            var textSpan = range == null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132601
+            var tokens = ComputeTokens(text.Lines, groupedSpans, tokenTypesToIndex);
+
+            return new LSP.SemanticTokens { ResultId = resultId, Data = tokens };
```
resultId is just passed in just to make this object.  just remove the resultId and the `new SemanticTokesn` here.  Just return `tokens`.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            // Since a TextSpan can have multiple ClassifiedSpans, all of the ClassifiedSpans except one (the token)
+            // should be a modifier.
+            var modifiers = textSpanToClassifiedSpans.Where(
+                s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
```
expensive linq allocation

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+                s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
+
+            // Filtering out the modifiers for now (will be added in step 5). We just want the primary token.
+            var tokenTypeClassifiedSpan = textSpanToClassifiedSpans.Except(modifiers).Single();
```
expensive linq allocation

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+        private static int GetTokenTypeIndex(ClassifiedSpan tokenTypeClassifiedSpan, Dictionary<string, int> tokenTypesToIndex)
+        {
+            s_classificationTypeToSemanticTokenTypeMap.TryGetValue(tokenTypeClassifiedSpan.ClassificationType, out var tokenTypeStr);
+            Debug.Assert(tokenTypeStr != null);
```
Contract.  this needs to blowup in this case.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            return tokenTypesToIndex[tokenTypeStr];
+        }
+
+        private static TokenModifiers GetTokenTypeModifierBits(IEnumerable<ClassifiedSpan> modifiers)
```
IEnumerable bad.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+        {
+            // Roslyn currently only has one modifier, but this logic should be updated if more are added
+            // in the future.
+            if (modifiers.IsEmpty())
```
Linq bad.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            }
+
+            Debug.Assert(modifiers.Single().ClassificationType == ClassificationTypeNames.StaticSymbol);
+            return TokenModifiers.Static;
```
we literally only have this case.  i would just check for this against the original classified span info.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs:
```diff
@@ -0,0 +1,58 @@
+            Contract.ThrowIfNull(request.TextDocument);
+            var requestId = _tokensCache.GetNextResultId();
+
+            // The results from the range handler should not be involved in caching since we don't
```
then why do we have a _tokensCache? :)

allisonchou replied:
Oops, I updated the comment to make it clearer that the range handler _uses_ the cache, but doesn't update it.

CyrusNajmabadi commented:
Done with apss.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,146 @@
+        private readonly int _maxCachesPerDoc = 5;
+
+        /// <summary>
+        /// The next resultId available to use.
```
```suggestion
        /// The next resultId available to use. Atomically incremented with Interlocked, so this doesn't need to be protected by _semaphore.
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,146 @@
+        /// <summary>
+        /// Maps a LSP token type to its respective index recognized by LSP.
+        /// </summary>
+        public static readonly Dictionary<string, int> TokenTypesToIndex;
```
nit:

```suggestion
        public static readonly Dictionary<string, int> TokenTypeToIndex;
```

In general since it's a single->to->single mapping, you don't pluralize the name.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,146 @@
+        {
+            // If the resultId of the semantic tokens is null, don't cache anything since we'll
+            // be unable to retrieve the results later.
+            if (tokens.ResultId == null)
```
should this be a contract check instead.  i.e. who is the caller of this that is passing in tokens that dont' have a result id?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                return newSemanticTokens;
+            }
+
+            // The Data property is always populated on the server side, so it should never be null.
```
comment doesn't really make sense.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+            }
+
+            // The Data property is always populated on the server side, so it should never be null.
+            Contract.ThrowIfNull(oldSemanticTokensData);
```
enh......... you literally check it for null one statement above.  seems really unnecessary.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                ResultId = resultId
+            };
+
+            return edits;
```
def love how much cleaner this all got :)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+            // is changed, the entire token is replaced. We do this instead of directly comparing each
+            // value in the token individually so that we can potentially save on computation costs, since
+            // we can return early if we find that one value in the token doesn't match. However, there
+            // are trade-offs since our insertions/deletions are usually larger.
```
great comment.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                {
+                    case EditKind.Insert:
+                        indexToEditKinds.TryGetValue(edit.NewIndex, out var editKindWithoutInsert);
+                        indexToEditKinds[edit.NewIndex] = editKindWithoutInsert |= SemanticTokenEditKind.Insert;
```
interesting.  so SmenaticTokenEdit "ors" teh values together?  i.e. you could have an edit at an index that is both an insert and a delete?  do you have tests for this?

allisonchou replied:
Yep, the semantic token edits tests have scenarios covering all three possible cases (insert, delete, and update).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        indexToEditKinds[edit.OldIndex] = editKindWithoutDelete |= SemanticTokenEditKind.Delete;
+                        break;
+                    default:
+                        break;
```
nit:

either remove, or potentially have `// intentionally not handled`

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                    case EditKind.Delete:
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        indexToEditKinds[edit.OldIndex] = editKindWithoutDelete |= SemanticTokenEditKind.Delete;
+                        break;
```
so, it feels liek this is saying each index in this dictionar can be in one of three states:

1. insert
2. delete
3. insert + delete

And you can only transition from `insert->(insert + delete)` or `delete->(insert + delete)`.   i.e. you should never get two inserts at the same index, or two deletes at the same index.  Is that correct?

CyrusNajmabadi replied:
also, would it make sense to ahve a specific SemanticTokenEditKind.Update flag to represent case 3?

allisonchou replied:
Yep, there should never be two inserts or deletes at the same index.
And yes, that would make sense - I created a separate Update flag in the latest round of revisions. Thanks!

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+            // may have both an insertion and deletion, in which case we can combine the two into a
+            // single update. We use the dictionary below to keep track of whether an index contains
+            // an insertion, deletion, or both.
+            var indexToEditKinds = new Dictionary<int, SemanticTokenEditKind>();
```
use pooled dictionary.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else
+                {
```
add a contract call that it has the delete flag.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+
+            // We sort the edit kinds by index since we need to know what kind of edits surround
+            // a given index in order to potentially combine them into one edit.
+            var editIndexes = indexToEditKinds.Keys.ToArray();
```
Indices is plural of Index

CyrusNajmabadi replied:
so these keys can be in a random order (i.e. dictionaries don't ensure a specific key ordering). is your algorithm correct in the face of that?  should you sort these indices first so we process things in an increasing order?

JoeRobich replied:
Just a note that indexes and indices are both accepted plural forms of index. Indices, however, is preferred in mathematical or technical fields. https://www.merriam-webster.com/dictionary/index

CyrusNajmabadi replied:
https://media2.giphy.com/media/3ohzdQChVBWTRSkILe/giphy.gif

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+            var editIndexes = indexToEditKinds.Keys.ToArray();
+            Array.Sort(editIndexes);
+
+            for (var i = 0; i < editIndexes.Length; i++)
```
having two "indices" here (one called 'i' and one called 'initialEditIndex') is pretty confusing.  but i'm struggling to find good names here .

allisonchou replied:
I changed 'i' to 'orderedEditNumber' and 'initialEditIndex' to 'orderedTokenNumber' for (hopefully) greater clarity. I also added an [example](https://github.com/dotnet/roslyn/pull/46407/files#diff-2cba795c77ec52f04332b943553ec1c3R155) in a comment.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert) && initialEditKind.HasFlag(SemanticTokenEditKind.Delete))
+                {
+                    i = AddInsertionDeletionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
```
so initialEditIndex is only passed in for two reasons:

1. to get the grouped semantic token.
2. to determine the start position of the final edit.  

Consider instead passing in those separate things as they will be easier to name and likely clearer.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+            {
+                var deleteCount = 5;
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(newGroupedSemanticTokens[initialEditIndex].ToArray());
```
instead of effectively allocating each token into an arrray, instead, have a method on the token that says "please add yourself to this ArrayBuilder at thsi position.  then we're alloc free here.

allisonchou replied:
I don't know if I fully understand this - maybe the name of the `ToArray` method is possibly confusing, since it's actually a method we define? I renamed it to `ConvertToArray`. But we're only converting one overall token here.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                // each other in the file.
+                while (i + 1 < editIndexes.Length && editIndexes[i + 1] == editIndexes[i] + 1)
+                {
+                    var currentEditKind = indexToEditKinds[editIndexes[i + 1]];
```
i woudl call this `nextEditKind`, because you're literally looking at the *next* thing.  The 'curent' is the one you already had.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                // An "update" (i.e. a dual insertion/deletion) can be combined with an update,
+                // deletion, or insertion that directly follows it. If combined with an insertion
+                // or deletion, only that type is allowed in the combined edit afterwards.
+                var editKind = SemanticTokenEditKind.None;
```
starting this with 'none' feels weird. don't you know what you started with at 'i' (i.e. a delete/insert/update)?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: initialEditIndex * 5, deleteCount: deleteCount, data: Array.Empty<int>()));
+                return i;
+            }
```
my original thought was that you could add these three methods together.  however, it's prob nice to keep them separate and nice and clean.

That said, the methods for insert/delete are really simple and easy to understand.  have you considered just tighting the rule to "you can only merge a sequence of all-updates, a sequence of all-deletes, or a seqeunce of all-inserts"?

That would simplify the first method greatly.  yes, we wouldnt' merge a couple of cases (like update-update-update-insert-insert-insert).   However, those would still get broken into sequences that would get merged.  So that seems fine.

allisonchou replied:
That's a good point, the logic would be definitely simpler if we tightened the rules. I have updated the code to reflect your suggestion. Thank you!

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+        /// <summary>
+        /// Stores the values that make up the LSP representation of an individual semantic token.
+        /// </summary>
+        private readonly struct SemanticToken
```
make IEquatable<SemanticToken>

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+                       _tokenModifiers == token._tokenModifiers;
+            }
+
+            public override int GetHashCode()
```
remove.  you never hash these.  and move .Equals(object) and replace with .Equals(SemanticToken)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,231 @@
+            return ComputeTokens(text.Lines, groupedSpans, tokenTypesToIndex);
+        }
+
+        private static int[] ComputeTokens(
```
skipping this since it still uses linq/grouping.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,377 @@
+        {
+            None = 0,
+            Insert = 1,
+            Delete = 2,
```
nit.  you mark this as flags (which is good).  but i still recommend writing this as: 
```suggestion
            Delete = 1 << 1,
```

the bitshift helps make it very clear that these are bit values (And thus intended to be 'or-ed').  That said, as per an above discussion, i think these coudl be non-flags, and just have 3 values (Insert/Delete/Update).

allisonchou replied:
Yep, just added an Update value per your suggestion. I kept the None value for a total of 4 values since it's used for some logic in the code.

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensTests.cs:
```diff
@@ -0,0 +1,144 @@
+
+            Assert.Equal(expectedWholeDocResults2.Data, wholeDocResults2.Data);
+            Assert.Equal(expectedWholeDocResults2.ResultId, wholeDocResults2.ResultId);
+        }
```
i may have missed it.  but i would love some tests that don't actually ever check the raw values.  instead tehy do this:

// pseudo code:
`ApplyEdits(FullClassification(oldCode), SomeUserEdits) equals FullClassification(oldCode + someUserEdits)`

allisonchou replied:
Yep! I added three of these while addressing the last iteration of your feedback, there is this [test](https://github.com/dotnet/roslyn/pull/46407/files#diff-18039c1c8975d6f3dff5f170f34daedbR182) and the two tests below it that verify the above scenario.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs:
```diff
@@ -13,6 +13,9 @@ public static class ClassificationTypeNames
         /// Additive classifications types supply additional context to other classifications.
         /// </summary>
+        /// <remarks>
+        /// NOTE: Any updates to this class should also be reflected in LSP.
```
i'd recommend giving a full path to where the updates would need to happen.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,357 @@
+                    case EditKind.Insert:
+                        indexToEditKinds.TryGetValue(edit.NewIndex, out var editKindWithoutInsert);
+                        indexToEditKinds[edit.NewIndex] = editKindWithoutInsert == default ? SemanticTokenEditKind.Insert : SemanticTokenEditKind.Update;
+                        break;
```
1. what happens if we have two inserts at the same position.  it becomes an update.  is that ok?  same with multiple deletes

CyrusNajmabadi replied:
if this can't happen, then we should have a contract call that verifies that.

allisonchou replied:
I don't think it is possible to have two inserts or deletes at the same position, so I added a contract check. Thanks for the suggestion!

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,357 @@
+            {
+                var deleteCount = 5;
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(groupedSemanticToken.ConvertToArray());
```
still allocating :)  do:

groupedSemanticToken.AddToEnd(tokensToInsert)

and have AddToEnd do the adding in a non-allocating fashion.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,357 @@
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int i,
```
i => orderedEditNumber

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,357 @@
+                {
+                    deleteCount += 5;
+                    i++;
+                }
```
suggestion.  this walk is repeated in all three methods.  consider a helper that just takes in the array, and the type of edit kidn to group and it walks forward and tells you how much to skip forward.  you then call that, and use the result to inform your computations (i.e. increasing 'i' or upping your deletion count).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,249 @@
+            // A TextSpan can be associated with multiple ClassifiedSpans (i.e. if  a token has
+            // modifiers). We perform this grouping since LSP requires that each token is
+            // reported together with all its modifiers.
+            using var _1 = PooledDictionary<TextSpan, List<string>>.GetInstance(out var textSpanToClassificationTypes);
```
likely not ok.  seems like you'll still be allocating those lsits.  let's chat.

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs:
```diff
@@ -0,0 +1,307 @@
+            var rawTokens = await RunGetSemanticTokensAsync(workspace.CurrentSolution, locations["caret"].First());
+
+            Assert.True(Enumerable.SequenceEqual(rawTokens.Data, editsToTokens));
+        }
```
test name is TestConvertSemanticTokenEditsIntoSemanticTokens_InsertNewlineInMiddleOfFile.  but i'm not seeing how a newline was inserted in teh middle of the file.  can you help explain the test for me?

allisonchou replied:
The original markup can be found [here](https://github.com/dotnet/roslyn/pull/46407/files#diff-18039c1c8975d6f3dff5f170f34daedbR24) and is:
```C#
// Comment
static class C { }  
```

The updated markup that we test against is:
```C#
// Comment

static class C { }
```

CyrusNajmabadi replied:
yeah. don't love this :-/

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs:
```diff
@@ -0,0 +1,136 @@
+
+        static SemanticTokensCache()
+        {
+            // Computes the mapping between a LSP token type and its respective index recognized by LSP.
```
can you clarify "respective index recognized by lsp".  i'm not quite sure what that means.

allisonchou replied:
I modified the comment, I agree it was confusing before. Hopefully this is clearer:
```
        /// Maps a LSP token type to the index LSP associates with the token.
       /// This is needed since we need to report tokens back to LSP as ints,
       /// and LSP needs a way to decipher them.
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                {
+                    case EditKind.Insert:
+                        indexToEditKinds.TryGetValue(edit.NewIndex, out var editKindWithoutInsert);
+                        Contract.ThrowIfTrue(editKindWithoutInsert == SemanticTokenEditKind.Insert); // We shouldn't have two inserts at the same position
```
👍

CyrusNajmabadi replied:
is it ok to get an insert and an update at the same position?  seems like no, since an update is an insert+delete.  so that would still mean tw inserts at the same location.  or am i incorrect?

allisonchou replied:
You're right - I'll add a Contract check to make sure we can't get an insert/update and delete/update at the same position. Thanks!

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                        break;
+                    case EditKind.Delete:
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        Contract.ThrowIfTrue(editKindWithoutDelete == SemanticTokenEditKind.Delete); // We shouldn't have two deletions at the same position
```
note: Contract.Throw takes a message string.  so instead of comment do:

```suggestion
                        Contract.ThrowIfTrue(editKindWithoutDelete == SemanticTokenEditKind.Delete, "We shouldn't have two deletions at the same position"); 
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+            // orderedTokenNumber:          0         1         2
+            for (var orderedEditNumber = 0; orderedEditNumber < editIndices.Length; orderedEditNumber++)
+            {
+                var orderedTokenNumber = editIndices[orderedEditNumber];
+                var initialEditKind = indexToEditKinds[orderedTokenNumber];
```
```suggestion
                var orderedTokenNumber = editIndices[orderedEditNumber];
                var initialEditKind = indexToEditKinds[orderedTokenNumber];
                var editStartPoistion = orderedTokenNumber * 5;
```

(and then just use this below in all cases).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int orderedEditNumber,
```
consider: startEditIndex.  then name the one below `updatedOrderedEditNumber  -> endEditIndex`.  


start/end have pretty well understood semantics.  it also makes it clear why you'r edoing `endEditIndex - startEditIndex`

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                var updatedOrderedEditNumber = GetUpdatedOrderedEditNumber(
+                    SemanticTokenEditKind.Update, indexToEditKinds, editIndices, orderedEditNumber);
+
+                var deleteCount = 5 + 5 * (updatedOrderedEditNumber - orderedEditNumber);
```
nit:

```suggestion
                var deleteCount = 5 * (1 + (updatedOrderedEditNumber - orderedEditNumber));
```
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+            // Non-grouped semantic tokens: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14
+            // orderedEditNumber:           0                   1
+            // orderedTokenNumber:          0         1         2
+            for (var orderedEditNumber = 0; orderedEditNumber < editIndices.Length; orderedEditNumber++)
```
`currentEditIndex`?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                for (var i = 1; i <= updatedOrderedEditNumber - orderedEditNumber; i++)
+                {
+                    newGroupedSemanticTokens[editIndices[orderedEditNumber + i]].AddToEnd(tokensToInsert);
+                }
```
question.  if you started from`i = 0` would this be simpler as you could remove the `groupedSemanticToken.AddToEnd(tokensToInsert)` line at the top of hte method?  (you could even then not pass in groupedSemanticToken).  Or am i misreading this?

allisonchou replied:
Nope, you're right. I think I was thinking of your suggestion at one point, but it involve re-looking up things in arrays, which maybe is fine because it's inexpensive. I'll change it.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                int editStartPosition)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                groupedSemanticToken.AddToEnd(tokensToInsert);
```
same question about not passing this in, and just having the below loop start at 0.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                var updatedOrderedEditNumber = GetUpdatedOrderedEditNumber(
+                    SemanticTokenEditKind.Delete, indexToEditKinds, editIndices, orderedEditNumber);
+
+                var deleteCount = 5 + 5 * (updatedOrderedEditNumber - orderedEditNumber);
```
nit:

```suggestion
                var deleteCount = 5 * (1+ (updatedOrderedEditNumber - orderedEditNumber));
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+            }
+
+            // Returns the updated ordered edit number after we know how many edits we can combine.
+            static int GetUpdatedOrderedEditNumber(
```
GetEndIndexOfSameEditKind?  (not sure if that's better).  but i do like saying "End"

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+            public static IEnumerable<SequenceEdit> GetEdits(
+                SemanticToken[] oldSemanticTokens, SemanticToken[] newSemanticTokens)
+                => s_instance.GetEdits(oldSemanticTokens, oldSemanticTokens.Length, newSemanticTokens, newSemanticTokens.Length);
+        }
```
nit: move the nested types to other files.  (up to you).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                _tokenModifiers = tokenModifiers;
+            }
+
+            public void AddToEnd(ArrayBuilder<int> tokensToInsert)
```
👍

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+                tokensToInsert.Add(_tokenModifiers);
+            }
+
+            public bool Equals([AllowNull] SemanticToken otherToken)
```
```suggestion
            public bool Equals(SemanticToken otherToken)
```

(structs can't be null).  it's a tooling issue that we spit that out.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:
```diff
@@ -0,0 +1,376 @@
+
+        private enum SemanticTokenEditKind
+        {
+            None = 0,
```
nit: you have a 0 value, but check for `default` in some comparisons.  can you check against None instead.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+                [ClassificationTypeNames.RegexComment] = LSP.SemanticTokenTypes.Regexp, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.RegexGrouping] = LSP.SemanticTokenTypes.Regexp, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.RegexOtherEscape] = LSP.SemanticTokenTypes.Regexp, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.RegexQuantifier] = LSP.SemanticTokenTypes.Regexp, // TO-DO: Potentially change to custom type
```
this surprises me.  In typescript there is some amount of regexp classification:

![image](https://user-images.githubusercontent.com/4564579/90663870-0f6e0500-e1ff-11ea-8cee-34475b74541c.png)

So it feels like LSP must have some way of distinguishing at least some regexp info... can you look into that later?

allisonchou replied:
That's super interesting, I'll definitely look into this as a follow-up after this PR is merged. David mentioned that Razor is hardcoding mappings... not sure if TS is doing something similar?

JoeRobich replied:
Without custom tokens there are no other Regex types. Typescript in VSCode doesn't emit semantic tokens for regex, so the editor shows classifications from the textmate language.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            ClassifiedSpan[] classifiedSpans,
+            Dictionary<string, int> tokenTypesToIndex)
+        {
+            using var _ = ArrayBuilder<int>.GetInstance(out var data);
```
```suggestion
            using var _ = ArrayBuilder<int>.GetInstance(classifiedSpans.Count, out var data);
```

this will help when there are a lot of spans so we don't need a lot of resizes.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            ClassifiedSpan[] classifiedSpans,
+            int currentClassifiedSpanIndex,
+            Dictionary<string, int> tokenTypesToIndex,
+            // Out params
```
```suggestion
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+
+            // 1. Token line number delta, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            Contract.ThrowIfTrue(deltaLine < 0);
```
i reommend including messges.  that way if there are multiple contract calls we know which one failed.  otherwise we just get "Unexpected false in ComputeNextToken" and we don't know which was at fault (since there aren't line numbers in release).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            var originalTextSpan = classifiedSpan.TextSpan;
+
+            // Classified spans with the same text span should be combined into one token.
+            while (classifiedSpans[currentClassifiedSpanIndex].TextSpan == originalTextSpan)
```
```suggestion
            while (currentClassifiedSpanIndex < classifiedSpans.Length && classifiedSpans[currentClassifiedSpanIndex].TextSpan == originalTextSpan)
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+                    // 5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+                    modifierBits = TokenModifiers.Static;
+                }
+
+                if (currentClassifiedSpanIndex + 1 >= classifiedSpans.Length || classifiedSpans[currentClassifiedSpanIndex + 1].TextSpan != originalTextSpan)
+                {
+                    break;
+                }
```
```suggestion
```

allisonchou replied:
Hmm.. I looked into this, but I don't know if the logic is equivalent, since we only want to increment `currentClassifiedSpanIndex` in certain scenarios.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -0,0 +1,243 @@
+            s_classificationTypeToSemanticTokenTypeMap.TryGetValue(classificationType, out var tokenTypeStr);
+            Contract.ThrowIfNull(tokenTypeStr);
+
+            if (!tokenTypesToIndex.TryGetValue(tokenTypeStr, out var tokenTypeIndex))
+            {
+                Contract.Fail("Token type index not found.");
+            }
```
```suggestion
            Contract.ThrowIfFalse(tokenTypesToIndex.TryGetValue(tokenTypeStr, out var tokenTypeIndex));
```

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs:
```diff
@@ -0,0 +1,307 @@
+                0,        6,     1,    SemanticTokensCache.TokenTypeToIndex[LSP.SemanticTokenTypes.Class],    (int)TokenModifiers.Static, // 'C'
+                0,        2,     1,    SemanticTokensCache.TokenTypeToIndex[LSP.SemanticTokenTypes.Operator], 0, // '{'
+                0,        2,     1,    SemanticTokensCache.TokenTypeToIndex[LSP.SemanticTokenTypes.Operator], 0, // '}'
+            };
```
i don't love this.  i would recommend this actually be generated from a true call to classify that original markup  i.e. your test class can have:

```c#
private static readonly string s_standardCase = @"// Comment
...";
```

and you can refer to that in your tests and get this int-array form by calling your own classification methods.  i don't think you should hardcode this in in this form.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/Classifier.cs:
```diff
@@ -28,6 +28,9 @@ public static class Classifier
         }
 
+        /// <summary>
+        /// Returns classified spans in order.
```
```
        /// Returns classified spans in ascending <see cref="ClassifiedSpan.Span"/> order.  Classified spans will not overlap.  However, contiguous <see cref="ClassifiedSpan"/>s may have the same <see cref="ClassifiedSpan.Span"/>.  This occurs when there are multiple <see cref="ClassifiedSpan.ClassificaitonType"/>s for the same region of code.  For example, a reference to a static-method will have two spans, one that designates it as a method, and one that designates it as static.
```
